Pod/Classes/PINAnimatedImage.m

```diff
@@ -0,0 +1,494 @@
+  return nil;
+}
+
+- (CFTimeInterval)durationAtIndex:(NSUInteger)index
+{
+  return self.durations[index];
+}
+
++ (CGImageRef)imageAtIndex:(NSUInteger)index inMemoryMap:(NSData *)memoryMap width:(UInt32)width height:(UInt32)height bitmapInfo:(CGBitmapInfo)bitmapInfo
+{
+  if (memoryMap == nil) {
+    return nil;
+  }
+  
+  Float32 outDuration;
+  
+  size_t imageLength = width * height * kPINAnimatedImageComponentsPerPixel;
+  
+  //frame duration + previous images
+  NSUInteger offset = sizeof(UInt32) + (index * (imageLength + sizeof(outDuration)));
+  
+  [memoryMap getBytes:&outDuration range:NSMakeRange(offset, sizeof(outDuration))];
+  
+  BytePtr imageData = (BytePtr)[memoryMap bytes];
+  imageData += offset + sizeof(outDuration);
+  
+  NSAssert(offset + sizeof(outDuration) + imageLength <= memoryMap.length, @"Requesting frame beyond data bounds");
+  
+  //retain the memory map, it will be released when releaseData is called
+  CFRetain((CFDataRef)memoryMap);
+  CGDataProviderRef dataProvider = CGDataProviderCreateWithData((void *)memoryMap, imageData, width * height * kPINAnimatedImageComponentsPerPixel, releaseData);
+  CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
+  CGImageRef imageRef = CGImageCreate(width,
+                                      height,
+                                      kPINAnimatedImageBitsPerComponent,
+                                      kPINAnimatedImageBitsPerComponent * kPINAnimatedImageComponentsPerPixel,
+                                      kPINAnimatedImageComponentsPerPixel * width,
+                                      colorSpace,
+                                      bitmapInfo,
+                                      dataProvider,
+                                      NULL,
+                                      NO,
+                                      kCGRenderingIntentDefault);
+  CFAutorelease(imageRef);
+  
+  CGColorSpaceRelease(colorSpace);
+  CGDataProviderRelease(dataProvider);
+  
+  return imageRef;
+}
+
++ (UInt32)widthFromMemoryMap:(NSData *)memoryMap
+{
+  UInt32 width;
+  [memoryMap getBytes:&width range:NSMakeRange(2, sizeof(width))];
+  return width;
+}
+
++ (UInt32)heightFromMemoryMap:(NSData *)memoryMap
+{
+  UInt32 height;
+  [memoryMap getBytes:&height range:NSMakeRange(6, sizeof(height))];
+  return height;
+}
+
++ (UInt32)loopCountFromMemoryMap:(NSData *)memoryMap
+{
+  UInt32 loopCount;
+  [memoryMap getBytes:&loopCount range:NSMakeRange(10, sizeof(loopCount))];
+  return loopCount;
+}
+
++ (UInt32)frameCountFromMemoryMap:(NSData *)memoryMap
+{
+  UInt32 frameCount;
+  [memoryMap getBytes:&frameCount range:NSMakeRange(14, sizeof(frameCount))];
+  return frameCount;
+}
+
++ (Float32 *)createDurationsFromMemoryMap:(NSData *)memoryMap frameCount:(UInt32)frameCount frameSize:(NSUInteger)frameSize totalDuration:(CFTimeInterval *)totalDuration
+{
+  *totalDuration = 0;
+  Float32 *durations = (Float32 *)malloc(sizeof(Float32) * frameCount);
```

Comment:
Since we are dealing with C mallocing and what not, would it make more sense to have the caller of this method provide a buffer instead of sending them back something that was malloc'ed as a side effect? That seems more C-like.